### DISCUSSION TITLE
Allow operations in BPM field

### DISCUSSION THREAD

**Discussion by c-sig** [ppy/osu#26582] [Ideas]
In some software (Figma comes to mind), you're allowed to input operations that get solved and input. 
![image](https://github.com/ppy/osu/assets/61018516/22689754-2600-4ed6-9dd4-45f689574214)
![image](https://github.com/ppy/osu/assets/61018516/4af0909c-5fc3-4743-bf9e-6ae82526ef0f)

It would be a great QOL thing for the editor as sometimes a lot of timing changes regarding BPM has to be made when original timing is not technically correct (halving bpm or incorrect time signature). Saves a bit of time as you won't have to open a calculator

**Comment by OliBomby** [comment]
I think this is quite a nice feature. I also use it for all my text boxes in Mapping Tools.
The expression can be easily computed with `DataTable.Compute()` from `System.Data`.

**Reply by peppy** [reply]
I'm not against it.

**Reply by bdach** [reply]
> The expression can be easily computed with `DataTable.Compute()` from `System.Data`

please do not use that that sounds like `eval()` tier nonsense. i don't want to think about the amount of undefined behaviour or actual security holes that may stem from using that

**Reply by OliBomby** [reply]
> please do not use that that sounds like `eval()` tier nonsense. i don't want to think about the amount of undefined behaviour or actual security holes that may stem from using that

Isn't that a little harsh? `DataTable.Compute()` is meant for evaluating simple expressions on data tables. Using it with an empty data table is just evaluating simple equations. It can't execute arbitrary code. Since its part of the dotnet standard library I assume its not full of security holes. Do you prefer a nuget package to compute expressions, or a hand-crafted parser?

**Reply by bdach** [reply]
oh i do not think it is harsh. i will admit that prior to 7 hours ago i was not aware of the existence of the function you mention, but generally i profess the principle that 'there is no free lunch'. if you tell me that there is a 'simple function that can compute arbitrary arithmetic expressions', what i read is 'this here function calls into a subset of the language which is probably underspecified, probably has weird type semantics, and probably behaves weirdly under duress'.

so what follows is a fifteen-minute experiment on what this function can do. i won't mention that even *finding* a semblance of a spec of this DSL was difficult. but i did finally manage to produce it in https://learn.microsoft.com/en-us/dotnet/api/system.data.datacolumn.expression?view=net-6.0.

so, shall we?

```csharp
var dataTable = new DataTable();
Console.WriteLine(dataTable.Compute("2 + 'test'", "")); // "2test"
```

good start, js-style number and string concatenation.

```csharp
Console.WriteLine(dataTable.Compute("2 - 'test'", "")); // throws EvaluateException
```

well at least it doesn't return NaN i guess?

```csharp
Console.WriteLine(dataTable.Compute($"{int.MaxValue}", ""));          // 2147483647
Console.WriteLine(dataTable.Compute($"{int.MaxValue} + 1 - 1", ""));  // throws OverflowException
Console.WriteLine(dataTable.Compute($"{long.MaxValue}", ""));         // 9223372036854775807
Console.WriteLine(dataTable.Compute($"{long.MaxValue} + 1 - 1", "")); // throws OverflowException 
Console.WriteLine(dataTable.Compute($"{double.MaxValue}", ""));       // 1.7976931348623157E+308
Console.WriteLine(dataTable.Compute($"{double.MaxValue} * 2", ""));   // ∞
```

for something that is supposed to be a "simple arithmetic expression parser" this sure does appear to chain itself to c# types pretty hard.

more number shenanigans:

```csharp
Console.WriteLine(dataTable.Compute($"0.1 + 0.1", ""));                                                             // 0.2
Console.WriteLine(dataTable.Compute($"0.000000000000000000000000000001 + 0.000000000000000000000000000001", ""));   // 0.0000000000000000000000000000
Console.WriteLine(dataTable.Compute($"0.1 - 0.1", ""));                                                             // 0.0
Console.WriteLine(dataTable.Compute($"(0.000000000000000000000000000001 - 0.000000000000000000000000000001)", "")); // 0.0000000000000000000000000000
Console.WriteLine(dataTable.Compute($"(0.000000000000000000000000000001 * 0.000000000000000000000000000001)", "")); // 0
```

here we see the difference between decimal and double.

```csharp
Console.WriteLine(dataTable.Compute("9e999", ""));  // ∞
Console.WriteLine(dataTable.Compute("9e-999", "")); // 0
```

accepts arbitrary out-of-range scientific notation.

```csharp
Console.WriteLine(dataTable.Compute("'test' LIKE '%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%'", ""));
```

we have string functions!

but, *but*, the real kicker that makes me want to run away from this _very_ fast, is this:

```csharp
Console.WriteLine(dataTable.Compute("CONVERT(3 < 4, 'System.Int32')", ""));  // 1
Console.WriteLine(dataTable.Compute("CONVERT(3 < 4, 'System.Object')", "")); // True
Console.WriteLine(dataTable.Compute("CONVERT(45, 'System.StringComparison')", "")); 
// System.InvalidCastException: Invalid cast from 'System.Int32' to 'System.StringComparison'.
//    at System.Convert.DefaultToType(IConvertible value, Type targetType, IFormatProvider provider)
//    at System.Data.Common.SqlConvert.ChangeType2(Object value, StorageType stype, Type type, IFormatProvider formatProvider)
//    at System.Data.FunctionNode.EvalFunction(FunctionId id, Object[] argumentValues, DataRow row, DataRowVersion version)
//    at System.Data.FunctionNode.Eval(DataRow row, DataRowVersion version)
//    at System.Data.FunctionNode.Optimize()
//    at System.Data.DataExpression.Bind(DataTable table)
//    at Program.<Main>$(String[] args)
//    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
//    at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
```

yes we have a type conversion primitive! and one that accepts _arbitrary types_ because it's a thin wrapper over `Convert.ChangeType()`!

now to its credit, something like the following does not appear to work:

```csharp
namespace Test {
    class Exploit {
        static Exploit() {
            Console.WriteLine("uh oh jinkies");
        }
    }
}

Console.WriteLine(dataTable.Compute($"CONVERT(45, '{typeof(Test.Exploit).FullName}')", ""));
// System.Data.EvaluateException: Invalid type name 'Test.Exploit'.
//    at System.Data.FunctionNode.GetDataType(ExpressionNode node)
//    at System.Data.FunctionNode.Eval(DataRow row, DataRowVersion version)
//    at System.Data.FunctionNode.Optimize()
//    at System.Data.DataExpression..ctor(DataTable table, String expression, Type type)
//    at Program.<Main>$(String[] args)
//    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
//    at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
```

likely because [there does seem to be some type filtering involved in the internals](https://github.com/dotnet/runtime/blob/05fabacb4fc3cf34faf21819b6fcc9892d2484c0/src/libraries/System.Data.Common/src/System/Data/Filter/FunctionNode.cs#L299-L309), but only _after_ the type is retrieved - seems pretty dangerously close to arbitrary code execution doesn't it?

to finish off here's a few cheap crashes using the rest of the arsenal.

```csharp
Console.WriteLine(dataTable.Compute("sum([test])", "")); // Cannot find column [test].
Console.WriteLine(dataTable.Compute("substring('hello', 1, -17)", "")); // Substring() argument is out of range.
```

i personally do not think we should allow anything like this, i think it is out of scope of the game to be a calculator. but _if_ we are to do that, then _please_, let's make sure that at least the implementation of this is pretty well-constrained to inputs that can be considered sane.

**Reply by OliBomby** [reply]
Thanks for the extensive research. The output of this method does seem to stay sane, but I see your concern of it doing some out-of-line stuff behind the scenes if you input some weird stuff.

**Reply by Qunas101** [reply]
DataTable.Compute() sounds a bit over engineered for a simple task of doubling/halving a BPM. Supporting just the * / + - signs should be plenty and a simple hand made parser could be fine
